connectivity/drivers/ble/FEATURE_BLE/COMPONENT_BlueNRG_2/BlueNrg2HCIDriver.cpp

```diff
@@ -0,0 +1,631 @@
+            hciCoreReadMaxDataLen();
+        }
+    }
+
+    void hciCoreReadMaxDataLen(void)
+    {
+        /* if LE Data Packet Length Extensions is supported by Controller and included */
+        if ((hciCoreCb.leSupFeat & HCI_LE_SUP_FEAT_DATA_LEN_EXT) &&
+                (hciLeSupFeatCfg & HCI_LE_SUP_FEAT_DATA_LEN_EXT)) {
+            /* send next command in sequence */
+            HciLeReadMaxDataLen();
+        } else {
+            /* send next command in sequence */
+            HciLeRandCmd();
+        }
+    }
+
+    void bluenrg_2_reset()
+    {
+        /* Reset BlueNRG_2 SPI interface. Hold reset line to 0 for 1500us */
+        rst = 0;
+        wait_us(1500);
+        rst = 1;
+
+        /* Wait for the radio to come back up */
+        wait_us(100000);
+    }
+
+    mbed::DigitalOut rst;
+    bool reset_received;
+    bool bluenrg_2_initialized;
+    bool enable_link_layer_mode_ongoing;
+};
+
+/**
+ * Transport driver of the ST BlueNRG_2 shield.
+ * @important: With that driver, it is assumed that the SPI bus used is not shared
+ * with other SPI peripherals. The reasons behind this choice are simplicity and
+ * performance:
+ *   - Reading from the peripheral SPI can be challenging especially if other
+ *     threads access the same SPI bus. Indeed it is common that the function
+ *     spiRead yield nothings even if the chip has signaled data with the irq
+ *     line. Sharing would make the situation worse and increase the risk of
+ *     timeout of HCI commands / response.
+ *   - This driver can be used even if the RTOS is disabled or not present it may
+ *     may be usefull for some targets.
+ *
+ * If The SPI is shared with other peripherals then the best option would be to
+ * handle SPI read in a real time thread woken up by an event flag.
+ *
+ * Other mechanisms might also be added in the future to handle data read as an
+ * event from the stack. This might not be the best solution for all BLE chip;
+ * especially this one.
+ */
+class TransportDriver : public CordioHCITransportDriver {
+public:
+    /**
+     * Construct the transport driver required by a BlueNRG_2 module.
+     * @param mosi Pin of the SPI mosi
+     * @param miso Pin of the SPI miso
+     * @param sclk Pin of the SPI clock
+     * @param irq Pin used by the module to signal data are available.
+     */
+    TransportDriver(PinName mosi, PinName miso, PinName sclk, PinName ncs, PinName irq)
+        : spi(mosi, miso, sclk), nCS(ncs), irq(irq), _spi_thread(osPriorityNormal, SPI_STACK_SIZE, _spi_thread_stack)
+    {
+        _spi_thread.start(mbed::callback(this, &TransportDriver::spi_read_cb));
+    }
+
+    virtual ~TransportDriver() { }
+
+    /**
+     * @see CordioHCITransportDriver::initialize
+     */
+    virtual void initialize()
+    {
+        // Setup the spi for 8 bit data, low clock polarity,
+        // 2-edge phase, with an 1MHz clock rate
+        spi.format(8, 1);
+        spi.frequency(1000000);
```

Comment:
Have you tried to operate the SPI at a faster clock rate ? The BlueNRG MS operates at 8MHz.

Reply:
No, this is a limitation of BlueNRG-2